java-checks/src/main/java/org/sonar/java/checks/NullPointerCheck.java

```diff
@@ -453,138 +433,97 @@ void registerAssignedSymbol(Symbol symbol) {
+      if (tree.is(Tree.Kind.METHOD_INVOCATION)) {
+        MethodInvocationTree methodInvocation = (MethodInvocationTree) tree;
+        Symbol symbol = methodInvocation.symbol();
+        if (symbol.isMethodSymbol()) {
+          MethodJavaSymbol methodSymbol = (MethodJavaSymbol) symbol;
+          List<JavaSymbol> parameters = methodSymbol.getParameters().scopeSymbols();
+          if (!parameters.isEmpty()) {
+            for (int i = 0; i < methodInvocation.arguments().size(); i += 1) {
+              // in case of varargs, there could be more arguments than parameters. in that case, pick the last parameter.
+              if (checkNullity(parameters.get(i < parameters.size() ? i : parameters.size() - 1)) == NullableState.NOTNULL) {
+                this.checkForIssue(methodInvocation.arguments().get(i),
+                  String.format("'%%s' is nullable here and method '%s' does not accept nullable argument", methodSymbol.name()),
+                  String.format("method '%s' does not accept nullable argument", methodSymbol.name()));
+              }
+            }
+          }
         }
       }
-      return AbstractValue.UNSET;
     }
 
-    // sets the value of the variable in the current state.
-    public void setVariableValue(VariableSymbol variable, AbstractValue value) {
-      variables.put(variable, value);
+    @Override
+    protected void onMemberAccess(ExecutionState executionState, MemberSelectExpressionTree tree) {
+      checkForIssue(tree.expression(), MESSAGE_NULLABLE_EXPRESSION, MESSAGE_NULL_LITERAL);
     }
 
     /**
-     * copies the value of each variables in fromState to this
+     * raises an issue if the passed tree can be null.
      *
-     * @param fromState state from which the values must be copied.
+     * @param tree tree that must be checked
+     * @param nullableMessage message to generate when the tree is an identifier or a method invocation
+     * @param nullMessage message to generate when the tree is the null literal
      */
-    public void copyValuesFrom(State fromState) {
-      for (VariableSymbol variable : fromState.variables.keySet()) {
-        this.setVariableValue(variable, fromState.getVariableValue(variable));
+    private void checkForIssue(Tree tree, String nullableMessage, String nullMessage) {
+      if (tree.is(Tree.Kind.IDENTIFIER)) {
+        Symbol symbol = ((IdentifierTree) tree).symbol();
+        if (checkNullity(tree) == NullableState.NULL) {
+          // prevents reporting issue multiple times
+          currentState.markValueAs(symbol, NullableState.UNKNOWN);
+          context.addIssue(tree, NullPointerCheck.this, String.format(nullableMessage, symbol.name()));
+        }
+      } else if (tree.is(Tree.Kind.METHOD_INVOCATION)) {
+        Symbol symbol = ((MethodInvocationTree) tree).symbol();
+        if (checkNullity(symbol) == NullableState.NULL) {
```

Comment:
![New](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/new.png) ![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Change this comparison to use the equals method. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1698)